pkg/kinds/general.go

```diff
@@ -0,0 +1,131 @@
+}
+
+// GrafanaResourceMetadata is standard k8s object metadata with helper functions
+type GrafanaResourceMetadata v1.ObjectMeta
+
+// GrafanaResource is a generic kubernetes resource with a helper for the common grafana metadata
+// This is a temporary solution until this object (or similar) can be moved to the app-sdk or kindsys
+type GrafanaResource[Spec interface{}, Status interface{}] struct {
+	APIVersion string `json:"apiVersion"`
+	Kind       string `json:"kind"`
+
+	Metadata GrafanaResourceMetadata `json:"metadata"`
+	Spec     *Spec                   `json:"spec,omitempty"`
+	Status   *Status                 `json:"status,omitempty"`
+
+	// Avoid extending
+	_ interface{}
+}
+
+// Annotation keys
+const annoKeyCreatedBy = "grafana.com/createdBy"
+const annoKeyUpdatedTimestamp = "grafana.com/updatedTimestamp"
+const annoKeyUpdatedBy = "grafana.com/updatedBy"
+
+// The folder identifier
+const annoKeyFolder = "grafana.com/folder"
+
+// Identify where values came from
+const annoKeyOriginName = "grafana.com/origin/name"
+const annoKeyOriginPath = "grafana.com/origin/path"
+const annoKeyOriginKey = "grafana.com/origin/key"
+const annoKeyOriginTime = "grafana.com/origin/time"
+
+func (m *GrafanaResourceMetadata) GetUpdatedTimestamp() *time.Time {
+	v, ok := m.Annotations[annoKeyUpdatedTimestamp]
+	if ok {
+		t, err := time.Parse(time.RFC3339, v)
+		if err != nil {
+			return &t
+		}
+	}
+	return nil
+}
+
+func (m *GrafanaResourceMetadata) SetUpdatedTimestamp(v *time.Time) {
+	if v == nil {
+		delete(m.Annotations, annoKeyUpdatedTimestamp)
+	} else {
+		m.Annotations[annoKeyUpdatedTimestamp] = v.Format(time.RFC3339)
+	}
+}
+
+func (m *GrafanaResourceMetadata) GetCreatedBy() string {
+	return m.Annotations[annoKeyCreatedBy]
+}
+
+func (m *GrafanaResourceMetadata) SetCreatedBy(user string) {
+	m.Annotations[annoKeyCreatedBy] = user // user GRN
+}
+
+func (m *GrafanaResourceMetadata) GetUpdatedBy() string {
+	return m.Annotations[annoKeyUpdatedBy]
+}
+
+func (m *GrafanaResourceMetadata) SetUpdatedBy(user string) {
+	m.Annotations[annoKeyUpdatedBy] = user // user GRN
+}
+
+func (m *GrafanaResourceMetadata) GetFolder() string {
+	return m.Annotations[annoKeyFolder]
+}
+
+func (m *GrafanaResourceMetadata) SetFolder(uid string) {
+	m.Annotations[annoKeyFolder] = uid
+}
+
+func (m *GrafanaResourceMetadata) SetOriginInfo(info *ResourceOriginInfo) {
+	delete(m.Annotations, annoKeyOriginName)
+	delete(m.Annotations, annoKeyOriginPath)
+	delete(m.Annotations, annoKeyOriginKey)
+	delete(m.Annotations, annoKeyOriginTime)
+	if info != nil || info.Name != "" {
+		m.Annotations[annoKeyOriginName] = info.Name
+		if info.Path != "" {
+			m.Annotations[annoKeyOriginPath] = info.Path
+		}
+		if info.Key != "" {
+			m.Annotations[annoKeyOriginKey] = info.Key
+		}
+		if info.Timestamp != nil {
+			m.Annotations[annoKeyOriginTime] = info.Timestamp.Format(time.RFC3339)
+		}
+	}
+}
+
+func (m *GrafanaResourceMetadata) GetOriginInfo(uid string) *ResourceOriginInfo {
```

Comment:
what are we intending to use uid for here?

Reply:
🤦   -- removed